### PR TITLE
[ci]Add test coverage using codecov [skip-ci]

### PR DESCRIPTION
**Pull request by AniqJaved**
# This Pull request:

## Checklist:

- [ ] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #

### CODE REVIEW

**Comment by AniqJaved** [comment]
Kindly run the ROOT CI TEST action in order to avoid conflicts.

**Comment by Axel-Naumann** [comment]
This was waiting for a team member to trigger the CI run - I did that just now.

**Review by Axel-Naumann — CHANGES_REQUESTED** [review]
Awesome, @AniqJaved - thanks a lot for all your work! 
- Could you please reduce the number of commits to something like 5 max? I'm happy to. show you how to rewrite the commit history.
- `.github/workflows/root-ci-config/__pycache__/build_utils.cpython-39.pyc`, `patch.txt`, and `.github/workflows/root-ci-config/test_file.py` won't be needed in the repo, please don't add them.
- `.github/workflows/test-coverage.yml` needs only one platform iiuc? Why does it have e.g. `build-macos`?

File: .github/workflows/test-coverage.yml [hunk 1]
```diff
@@ -0,0 +1,430 @@
+
+name: 'ROOT CI TEST'
```
Comment (Axel-Naumann) [inline .github/workflows/test-coverage.yml:2]: "Test Coverage"?

File: .github/workflows/test-coverage.yml [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+                    --head_ref       refs/pull/${{ github.event.pull_request.number }}/head:${{ github.event.pull_request.head.ref }}
+                    --repository     ${{ github.server_url }}/${{ github.repository }}
+              "
+
+      - name: Workflow dispatch
+        if:   github.event_name == 'workflow_dispatch'
+        run: ".github/workflows/root-ci-config/build_root.py
+                    --buildtype      ${{ matrix.config }}
+                    --platform       ${{ matrix.image }}
+                    --incremental    ${{ inputs.incremental }}
```
Comment (Axel-Naumann) [inline .github/workflows/test-coverage.yml:361]: See above.

File: .github/workflows/test-coverage.yml [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+      - name: Update artifacts after push to release branch
+        if:   github.event_name == 'push'
+        run: ".github/workflows/root-ci-config/build_root.py
+                    --buildtype      ${{ matrix.config }}
+                    --platform       ${{ matrix.image }}
+                    --incremental    false
+                    --base_ref       ${{ github.ref_name }}
+                    --repository     ${{ github.server_url }}/${{ github.repository }}
+              "
+
```
Comment (Axel-Naumann) [inline .github/workflows/test-coverage.yml:388]: Please remove.

File: .github/workflows/test-coverage.yml [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+                    --repository     ${{ github.server_url }}/${{ github.repository }}
+              "
+    
+      - name: Upload test results
+        if:   ${{ !cancelled() }}
+        uses: actions/upload-artifact@v3
+        with:
+          name:  Test Results ${{ matrix.image }} ${{ matrix.config }}
+          path: /github/home/ROOT-CI/build/TestResults.xml
+
```
Comment (Axel-Naumann) [inline .github/workflows/test-coverage.yml:395]: We don't need / want this.

File: .github/workflows/test-coverage.yml [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+      # - name: Setup tmate session
+      #   if: failure() && runner.debug == '1'
+      #   uses: mxschmitt/action-tmate@v3
+      #   with:
+      #     sudo: false
+      #   # When debugging increase to a suitable value!
+      #   timeout-minutes: ${{ github.event.pull_request && 40 || 20 }}
+
+      - name: Upload coverage to Codecov
+        if: ${{ matrix.image == 'fedora38-cov' && !cancelled() }}
```
Comment (Axel-Naumann) [inline .github/workflows/test-coverage.yml:411]:
```suggestion
        if: ${{ !cancelled() }}
```
As this file is now *only* about coverage we know this is a coverage build :-)

File: .github/workflows/test-coverage.yml [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+      #   with:
+      #     name: Test Coverage ${{ matrix.image }} ${{ matrix.config }}
+      #     path: /tmp/workspace/build/CoverageXML.xml
+      # - name: Setup tmate session
+      #   if: failure() && runner.debug == '1'
+      #   uses: mxschmitt/action-tmate@v3
+      #   with:
+      #     sudo: false
+      #   # When debugging increase to a suitable value!
+      #   timeout-minutes: ${{ github.event.pull_request && 40 || 20 }}
```
Comment (Axel-Naumann) [inline .github/workflows/test-coverage.yml:408]: Can this commented block go away? It doesn't seem to add much...

File: .github/workflows/test-coverage.yml [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+  event_file:
+    if: github.repository_owner == 'root-project' || github.event_name == 'pull_request'
+    name: "Upload Event Payload"
+    runs-on: ubuntu-latest
+    steps:
+    - name: Upload
+      uses: actions/upload-artifact@v3
+      with:
+        name: Event File
+        path: ${{ github.event_path }}
```
Comment (Axel-Naumann) [inline .github/workflows/test-coverage.yml:430]: I don't think we need this section. But we do want a newline at the end of the file :-)

File: .github/workflows/test-coverage.yml [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+      #   if: failure() && runner.debug == '1'
+      #   uses: mxschmitt/action-tmate@v3
+      #   with:
+      #     sudo: false
+      #   # When debugging increase to a suitable value!
+      #   timeout-minutes: ${{ github.event.pull_request && 40 || 20 }}
+
+      - name: Upload coverage to Codecov
+        if: ${{ matrix.image == 'fedora38-cov' && !cancelled() }}
+        uses: codecov/codecov-action@v3
```
Comment (Axel-Naumann) [inline .github/workflows/test-coverage.yml:412]: How is this authenticated? I suspect this needs a token? But that token won't be exposed to `pull_request` builds. Please have a look [here](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/), where they build with minimal credentials (`pull_request`), store the result as an artifact (we don't even need that!), and finally run the step that needs more permissions (in our case: to read secrets) to work just on the generated artifact. We can simply access that file in the `pull_request_target` part after the `pull_request` part building, testing, and generating the XML file has finished.

Let's discuss next week!

**Review by Axel-Naumann — CHANGES_REQUESTED** [review]
Very nice, several things we'll want to discuss next week, but it's an amazing first step!

**Pull request closed without merging**